test/integration/endpointslice/endpointslicemirroring_test.go

```diff
@@ -228,3 +231,213 @@ func TestEndpointSliceMirroring(t *testing.T) {
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-123",
+					Namespace: ns.Name,
+				},
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{{
+						Port: int32(80),
+					}},
+				},
+			}
+
+			customEndpoints := &corev1.Endpoints{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        "test-123",
+					Namespace:   ns.Name,
+					Labels:      map[string]string{},
+					Annotations: map[string]string{},
+				},
+				Subsets: []corev1.EndpointSubset{{
+					Ports: []corev1.EndpointPort{{
+						Port: 80,
+					}},
+					Addresses: []corev1.EndpointAddress{{
+						IP: "10.0.0.1",
+					}},
+				}},
+			}
+
+			_, err = client.CoreV1().Services(ns.Name).Create(context.TODO(), service, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("Error creating service: %v", err)
+			}
+
+			_, err = client.CoreV1().Endpoints(ns.Name).Create(context.TODO(), customEndpoints, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("Error creating endpoints: %v", err)
+			}
+
+			// Obtain ResourceVersion of the new endpoint slice created
+			var resVersion string
+			if err := wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				lSelector := discovery.LabelServiceName + "=" + service.Name
+				esList, err := client.DiscoveryV1beta1().EndpointSlices(ns.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+				if err != nil {
+					t.Logf("Error listing EndpointSlices: %v", err)
+					return false, err
+				}
+
+				if len(esList.Items) == 0 {
+					t.Logf("Waiting for EndpointSlice to be created")
+					return false, nil
+				}
+				endpointSlice := esList.Items[0]
+				resVersion = endpointSlice.ObjectMeta.ResourceVersion
+				return true, nil
+			}); err != nil {
+				t.Fatalf("endpoint slice not found: %v", err)
+			}
+
+			// update endpoint
+			tc.tweakEndpoint(customEndpoints)
+			_, err = client.CoreV1().Endpoints(ns.Name).Update(context.TODO(), customEndpoints, metav1.UpdateOptions{})
+			if err != nil {
+				t.Fatalf("Error updating endpoints: %v", err)
+			}
+
+			// verify the endpoint updates were mirrored
+			err = wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				lSelector := discovery.LabelServiceName + "=" + service.Name
+				esList, err := client.DiscoveryV1beta1().EndpointSlices(ns.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+				if err != nil {
+					t.Logf("Error listing EndpointSlices: %v", err)
+					return false, err
+				}
+
+				if len(esList.Items) == 0 {
+					t.Logf("Waiting for EndpointSlice to be created")
+					return false, nil
+				}
+				if len(esList.Items) > 1 {
+					return false, fmt.Errorf("Only expected 1 EndpointSlice, got %d", len(esList.Items))
+				}
+				endpointSlice := esList.Items[0]
+				if endpointSlice.ObjectMeta.ResourceVersion == resVersion {
+					t.Logf("Waiting for EndpointSlice to be updated")
+					return false, nil
+				}
```

Comment:
This feels like the wrong thing to check for. We should ensure that EndpointSlices mirrored for this Endpoints resource have these annotations. It shouldn't matter if it was the same one that was originally created, that feels more like an implementation detail. There's also a possibility that the resource version changed for some other reason here and we fail unexpectedly.

Reply:
this should be guaranteed by the test, that only tests Updates `TestEndpointSliceMirroringUpdates`, is an implementation detail to guarantee that updates are processed correctly (there is only worker in the controller to guarantee everything is serialized):
1. create endpoints without annotations or labels
2. get the resource version of the slice generated so we don't know when it was updated
3. annotate or label the endpoint and update
4. we need to wait for the updated endpoint slice, that will have a new resourcevesion because the endpointslicemirrorig controller will update the current one

Reply:
This is a tiny nit, but I think what we actually want is slightly different. Even though this test should pass 100% of the time given the clean/restricted environment it's running in, it would be possible for changes to the controller to break this. For example, controller logic could change in the future to replace the existing slice with a new one instead of updating the existing one or to mirror an endpoint with multiple slices. Instead I think this test should be ensuring that all relevant slices managed by the controller get the new annotation. Changes to the logic here seem unlikely, but I think it would be safer to not have the resource version + single resource expectations built into the test.

Reply:
fair enough